src/mame/drivers/jpmsru.cpp

```diff
@@ -1,64 +1,172 @@
 // license:BSD-3-Clause
-// copyright-holders:David Haywood
+// copyright-holders:David Haywood, SomeRandomGuyIdk
 /**********************************************************************
-
-    JPM S.R.U Hardware
-
-    The Stepper Reel Unit (SRU) was the first JPM hardware platform
-    to use CPU based technology as opposed to a purely mechanical
-    setup.
-
-    It really shows that this is the antecedent of System 80,
-    the sound hardware seems to be a basic discrete circuit (complete
-    with tone pot), and the hardware uses the older TMS9980A as a CPU.
-
-    TODO: Everything!, there are 10 different SRU subtypes, including
-    one purely for export, though this is a layout change rather than
-    anything significant.
-
+	
+	JPM Stepper Reel Unit
+	
+	JPM's first CPU-based fruit machine platform, from late 1978/1979.
+	Notably the first system to use stepper reels instead of EM reels.
+	Uses a 1.5MHz TMS9980A CPU together with some TTL for I/O providing
+	56 outputs (16 used by reels) & 24 inputs (8 used by optos), 
+	a ROM card holding game ROMs, and a selection of expansion boards. 
+	Sound is output by a simple 6-tone NE556-based circuit.
+	
+	TODO:
+	- Layouts
+	- Netlist audio works but isn't quite right. The tone pot needs to
+	  be set to 17% for it to not cut out, and even then popping can be heard.
+	  Needs a look from someone with more analog knowledge than me.
+	- Add remaining games
+	
+	Expansion boards:
+	Bus Extension
+	Optional board with 128 nibbles NVRAM and 24 DIP switches, only supported by early JPM games
+	
+	Input Extension
+	Same as Bus Extension except with 8 extra inputs replacing the NVRAM (moved to ROM card), used by club games
+	
+	Logic Extension
+	56 extra outputs addressed from CRU memory, accessed via 9 existing outputs
+	
+	Maxi Logic Extension
+	64 extra outputs addressed from CRU memory, accessed via 9 existing outputs
+	
+	Mini Logic Extension
+	16 extra outputs addressed directly from CRU memory
+	
+	Output Extension
+	16 extra outputs addressed from main memory
+	
+	ROM cards:
+	Most SRU games used a 3K ROM card for storage. A few later games had a 4K card,
+	and a 6K card with 512 nibbles of NVRAM was used for club games.
+	
 **********************************************************************/
 
 #include "emu.h"
+
 #include "cpu/tms9900/tms9980a.h"
+#include "machine/timer.h"
+#include "machine/steppers.h"
+#include "machine/genfruit.h"
+#include "machine/nvram.h"
+#include "video/awpvid.h"
+#include "machine/netlist.h"
+#include "netlist/nl_setup.h"
+#include "speaker.h"
+#include "audio/nl_jpmsru.h"
```

Comment:
These should be grouped with stuff from src/mame, src/devices and src/emu, and src/lib separated.  We order the groups most-dependent to least-dependent to help catch cases of headers not including other headers that they should.

Reply:
Fixed.